File: auth-service/src/main/java/com/eirsteir/coffeewithme/authservice/validation/PasswordConstraintValidator.java
package com.eirsteir.coffeewithme.authservice.validation;

import com.google.common.base.Joiner;
import java.util.Arrays;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import org.passay.*;

public class PasswordConstraintValidator implements ConstraintValidator<ValidPassword, String> {

  @Override
  public void initialize(final ValidPassword arg0) {}

  @Override
  public boolean isValid(final String password, ConstraintValidatorContext context) {

    PasswordValidator validator =
        new PasswordValidator(
            Arrays.asList(
                new LengthRule(8, 64), new QwertySequenceRule(3, false), new WhitespaceRule()));
    RuleResult result = validator.validate(new PasswordData(password));

    if (result.isValid()) return true;

    context.disableDefaultConstraintViolation();
    context
        .buildConstraintViolationWithTemplate(Joiner.on(",").join(validator.getMessages(result)))
        .addConstraintViolation();
    return false;
  }
}

File: auth-service/src/main/kotlin/com/eirsteir/coffeewithme/authservice/validation/PasswordConstraintValidator.kt
package com.eirsteir.coffeewithme.authservice.validation

import com.google.common.base.Joiner
import org.passay.*
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class PasswordConstraintValidator : ConstraintValidator<ValidPassword, String> {
    override fun initialize(arg0: ValidPassword) {}

    override fun isValid(password: String?, context: ConstraintValidatorContext): Boolean {
        val validator = PasswordValidator(
            listOf(
                LengthRule(8, 64), QwertySequenceRule(3, false), WhitespaceRule()
            )
        )
        val result = validator.validate(PasswordData(password))
        if (result.isValid)
            return true

        context.disableDefaultConstraintViolation()
        context
            .buildConstraintViolationWithTemplate(Joiner.on(",").join(validator.getMessages(result)))
            .addConstraintViolation()
        return false
    }
}